src/serde_helpers.rs

```diff
@@ -215,10 +215,16 @@ pub mod uuid_as_binary {
         D: Deserializer<'de>,
     {
         let binary = Binary::deserialize(deserializer)?;
-        if binary.subtype == BinarySubtype::Uuid {
+        if binary.subtype == BinarySubtype::Uuid || binary.subtype == BinarySubtype::UuidOld {
             if binary.bytes.len() == 16 {
                 let mut bytes = [0u8; 16];
                 bytes.copy_from_slice(&binary.bytes);
+
+                if binary.subtype == BinarySubtype::UuidOld {
+                    bytes[0..8].reverse();
+                    bytes[8..16].reverse();
+                }
```

Comment:
I'm not sure baking this assumption into the deserializer is the right approach.  The correct logic for this depends on what client created the data, which is why the new UUID subtype was created.  The logic here is only correct for data created by the Java client.  

Specifics on the correct deserialization rules are here: https://github.com/mongodb/mongo-python-driver/blob/master/bson/binary.py#L251-L262

Since the data format itself provides no way to know which format to use, this would have to be handled either by providing an argument to the deserializer itself (probably not a great idea), or deserializing to a target type that understands the appropriate byte order.  One alternative would be to create separate types for each format: 

```rust
struct PythonLegacyUuid(uuid::Uuid);
struct JavaLegacyUuid(uuid::Uuid);
struct CSharpLegacyUuid(uuid::Uuid);
```

and then impl Deref<Target=uuid::Uuid>` (& also DerefMut).

Another alternative would be to create a single LegacyUuid type that stores the bytes untouched, and allows the user to choose the appropriate ordering by method call.
```rust
struct LegacyUuid([u8;16]);
impl LegacyUuid {
    fn java_uuid(&self) -> Uuid { todo!() }
    fn python_uuid(&self) -> Uuid { todo!() }
    fn csharp_uuid(&self) -> Uuid { todo!() }
}
```

These types can be created outside of the bson create, but I would also love to see one or both alternatives within as well, since they solely exist to deal with legacy mongodb issues.

Reply:
Thank you for your suggestion. I will add `JavaLegacyUuid`, `PythonLegacyUuid`, and `CSharpLegacyUuid` as you suggested.